grove/measurements/estimation.py

```diff
@@ -0,0 +1,161 @@
+    in pauli_terms.  The bits in each element of bitstring_results correspond
+    to a numerical ordering of the maximal weight element of pauli_terms.
+
+    An example:
+
+    A maximum weight pauli term is Z1Z5Z6.
+
+    each element of bitstring_results is something in
+    :math:`\{0, 1\}^{\otimes 3}` where the first bit corresponds to the measured
+    value of qubit 1, the second bit corresponds to the measured value of qubit
+    5, and the third bit corresponds to the measured value of qubit 6.
+
+    :param pauli_terms:
+    :param bitstring_results:
+    :return:
+    """
+    max_weight_pauli_index = np.argmax(
+        [len(term.get_qubits()) for term in pauli_terms])
+    active_qubit_indices = sorted(
+        pauli_terms[max_weight_pauli_index]._ops.keys())
+    index_mapper = dict(zip(active_qubit_indices,
+                            range(len(active_qubit_indices))))
+
+    results = np.zeros((len(pauli_terms), len(bitstring_results)))
+
+    # convert to array so we can fancy index into it later.
+    # list() is needed to cast because we don't really want a map object
+    bitstring_results = list(map(np.array, bitstring_results))
+    for row_idx, term in enumerate(pauli_terms):
+        memory_index = np.array(list(map(lambda x: index_mapper[x],
+                                         sorted(term.get_qubits()))))
+
+        results[row_idx, :] = [-2 * (sum(x[memory_index]) % 2) + \
+                               1 for x in bitstring_results]
+    return results
+
+
+def estimate_pauli_set(pauli_term_set, basis_transform_dict, program,
+                       variance_bound, quantum_resource):
+    """
+    Estimate the mean of a sum of pauli terms to set variance
+
+    The sample variance is calculated by
+
+    .. math::
+        \begin{align}
+        \mathrm{Var}[\hat{\langle H \rangle}] = \sum_{i, j}h_{i}h_{j}
+        \mathrm{Cov}(\hat{\langle P_{i} \rangle}, \hat{\langle P_{j} \rangle})
+        \end{align}
+
+    :param pauli_term_set: list of pauli terms to measure simultaneously
+    :param basis_transform_dict: basis transform dictionary where the key is
+                                 the qubit index and the value is the basis to
+                                 rotate into. Valid basis is [I, X, Y, Z].
+    :param program: program generating a state to sample from
+    :param variance_bound:  Bound on the variance of the estimator for the
+                            PauliSum. Remember this is the SQUARE of the
+                            standard error!
+
+    :param quantum_resource: quantum abstract machine object
+    :return: estimated expected value, covariance matrix, variance of the
+             estimator, and the number of shots taken
+    """
+    if not isinstance(pauli_term_set, list):
+        raise TypeError("pauli_term_set needs to be a list")
+
+    # check if each term commutes with everything
+    if len(commuting_sets(sum(pauli_term_set))) != 1:
```

Comment:
How expensive is this call? For large sets of Pauli terms, it may be ok to make this check optional and depend on a valid input.